lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -4426,23 +4710,500 @@ class JVPEmitter final
     assert(differentialCall->getNumResults() == 1 &&
            "Expected differential to return one result");
 
-    // TODO: Generalize for indirect results, multiple results, etc.
-    auto origResult = ai->getResult(actualIndices.source);
+    // Get the original results of the `apply` instructions.
+    SmallVector<SILValue, 8> origDirResults;
+    collectAllExtractedElements(ai, origDirResults);
+    SmallVector<SILValue, 8> origAllResults;
+    collectAllActualResultsInTypeOrder(
+        ai, origDirResults, ai->getIndirectSILResults(), origAllResults);
+    auto origResult = origAllResults[actualIndices.source];
 
-    // Extract all direct results from the differential.
+    // Get the differential results of the `apply` instructions.
     SmallVector<SILValue, 8> differentialDirResults;
-    extractAllElements(differentialCall, diffBuilder, differentialDirResults);
-    // Get all differential results in type-defined order.
+    collectAllExtractedElements(differentialCall, differentialDirResults);
     SmallVector<SILValue, 8> differentialAllResults;
     collectAllActualResultsInTypeOrder(
         differentialCall, differentialDirResults,
         differentialCall->getIndirectSILResults(), differentialAllResults);
-    auto differentialResult = differentialAllResults[actualIndices.source];
+    auto differentialResult = differentialAllResults.front();
 
     // Add tangent for original result.
-    assert(actualIndices.source == 0 && "Expected result index to be first.");
-    setTangentValue(bb, origResult,
+    if (origResult->getType().isObject()) {
+      if (!origResult->getType().is<TupleType>()) {
+        setTangentValue(bb, origResult,
+            makeConcreteTangentValue(differentialResult));
+      } else {
+        unsigned numberOfDestructureTupleInsts = 0;
+        bool notSetValue = true;
+        for (auto use : ai->getUses()) {
+          if (auto *dti = dyn_cast<DestructureTupleInst>(use->getUser())) {
```

Comment:
Shouldn't you use the `getSingleDestructureTupleInstUser` helper?

Reply:
Ahh yup accidentally put some of the `There should only be one use of the tuple that is a 'destructure_tuple' inst` verification here as well, moved it to `getSingleDestructureTupleUser` and called that function here instead. Still kept the `should only have one active result from the tuple logic` here though.